src/Adapter/AmqpLib/Queue/AmqpQueue.php

```diff
@@ -0,0 +1,183 @@
+
+        try {
+            $amqplibChannel->basic_consume(
+                $this->getName(),
+                $tag,
+                false,
+                !$this->autoAck,
+                $this->definition->isExclusive(),
+                false,
+                function (AMQPMessage $message) use ($handler) {
+                    $receivedMessage = new AmqpReceivedMessage($this, $message);
+
+                    $handler($receivedMessage);
+                }
+            );
+            // Loop as long as the channel has callbacks registered
+            while ($amqplibChannel->is_consuming()) {
+                $amqplibChannel->wait(null, false, $this->getChannel()->getConnection()->getReadTimeout());
+            }
+        } catch (\Throwable $e) {
+            if ($amqplibChannel->is_consuming()) {
+                $this->cancelConsumer($tag);
+            }
+
+            if ($e instanceof AMQPTimeoutException || false !== \strpos(\strtolower($e->getMessage()), 'consumer timeout exceed')) {
+                throw new ConsumerTimeoutExceedException('Consumer timeout exceed.', 0, $e);
+            }
+
+            throw $e;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function cancelConsumer(string $tag): void
+    {
+        $this->channel->getChannel()->basic_cancel($tag);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function get(): ?ReceivedMessageInterface
+    {
+        $message = $this->channel->getChannel()->basic_get($this->getName());
+
+        if ($message) {
+            return new AmqpReceivedMessage($this, $message);
+        }
+
+        return null;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function purge(): void
+    {
+        $this->channel->getChannel()->queue_purge($this->getName());
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * Declares queue as a side-effect
+     */
+    public function countMessages(): int
+    {
+        return $this->declare()[1] ?? 0;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getName(): string
+    {
+        return $this->definition->getName();
+    }
+
+    /**
+     * @return array|null   declare result, includes queue size
+     */
+    public function declare(): ?array
+    {
+        $arguments = new AMQPTable();
+
+        foreach ($this->definition->getArguments() as $argument) {
+            $arguments->set($argument->getName(), $argument->getValue());
+        }
+
+        return $this->channel->getChannel()->queue_declare(
+            $this->getName(),
+            $this->definition->isPassive(),
+            $this->definition->isDurable(),
+            $this->definition->isExclusive(),
+            $this->definition->isAutoDelete(),
+            false,
+            $arguments
+        );
+    }
+
+    /**
+     * @param string $exchangeName
+     * @param string $routingKey
+     */
+    public function bind(string $exchangeName, string $routingKey)
```

Comment:
1. `void` return type.
2. This method does not declared in interface. Why must be here?

Reply:
because this is intended to be an analog to \AMQPQueue from PECL extension, php-amqplib does not provide such model